pep-0572.rst

```diff
@@ -275,16 +275,47 @@ i`` part establishes that ``i`` is local to the comprehension, but the
 ``i :=`` part insists that ``i`` is not local to the comprehension.
 The same reason makes these examples invalid too::
 
-    [[(j := j) for i in range(5)] for j in range(5)]
-    [i := 0 for i, j in stuff]
-    [i+1 for i in i := stuff]
+    [[(j := j) for i in range(5)] for j in range(5)] # compiler TargetScopeError
+    [i := 0 for i, j in stuff]                       # compiler TargetScopeError
+    [i+1 for i in (i := stuff)]                      # compiler TargetScopeError
+
+While it's technically possible to assign consistent semantics to these cases,
+it's difficult to determine whether those semantics actually make *sense* in the
+absence of real use cases. Accordingly, the reference implementation will ensure
+that such cases raise ``TargetScopeError``, a new subclass of ``SyntaxError``
+(with the same signature), rather than executing with implementation defined
+behaviour.
+
+This restriction applies even if the assignment expression is never executed::
+
+    [False and (i := 0) for i, j in stuff]     # compiler TargetScopeError
+    [i for i, j in stuff if True or (j := 1)]  # compiler TargetScopeError
+
+For the comprehension body (the part before the first "for" keyword) and the
+filter expression (the part after "if" and before any nested "for"), this
+restriction applies solely to target names that are also used as iteration
+variables in the comprehension. Lambda expressions appearing in these
+positions introduce a new explicit function scope, and hence may use assignment
+expressions with no additional restrictions.
+
+Due to design constraints in the reference implementation, named expressions
+are disallowed entirely as part of comprehension iterable expressions (the part
```

Comment:
Is "comprehension iterable expression" a technical term that's used elsewhere? It's quite a mouthful, and if you were to mention it to me without the explanation you add here I would not have known what exactly it meant.

Reply:
I don't think we've ever officially settled on a concise name for it.

The closest "comprehension iterable expression" gets to being official is that the reference docs at https://docs.python.org/3/reference/expressions.html#displays-for-lists-sets-and-dictionaries use the phrase "iterable expression" in a couple of places where they need to refer to what would be more explicitly described as "the `or_test` section of the `comp_for` grammar node".

Reply:
OK, so it refers to the part after 'in' at any level, not just at the top level.